### DISCUSSION TITLE
Rewording commits in normal mode and creating commits in insert mode?

### DISCUSSION THREAD

**Discussion by adamroyjones** [magit/magit#5047] [Support] [answered]
I use magit and evil-mode (including evil-collection). I am, however, a little clueless about something.

I was hoping to customise things so that:

- when creating a commit (magit-commit-create), the buffer is opened in insert mode; and
- when rewording a commit (magit-commit-reword), the buffer is opened in normal mode.

I've tried adding advice, e.g.

```elisp
(advice-add 'magit-commit-reword :after #'(lambda (&rest _) (evil-normal-state)))
```

and I can see that adding advice does lead to calls, but it's not doing what I'd like. It's like the call is happening too early.

I can't see any relevant hooks or modes that apply to one act and not the other.

All I can think of is trying to perform invasive surgery with :around advice that replaces rather a lot more—but that feels too heavy. I've searched everywhere and can't find a clear solution.

I'm sure someone more deft would have a better idea. Does anyone know the direction I should be looking in?

**Comment by tarsius** [comment, marked as answer]
`git-commit-setup-hook` would probably be a good place to enable the desired vi mode.

> Does anyone know the direction I should be looking in?

If the above does not work, you'll have to ask for help elsewhere, to reach follow Evil users.

**Reply by adamroyjones** [reply]
That was the key that unlocked it—thanks for guiding me!

For anyone wanting to do the same thing: the below sufficed. (I don't claim that it's perfect.)

```elisp
(defvar arj/magit-commit-or-reword nil)

(advice-add 'magit-commit-create :after #'(lambda (&rest _) (setq arj/magit-commit-or-reword :commit)))
(advice-add 'magit-commit-reword :after #'(lambda (&rest _) (setq arj/magit-commit-or-reword :reword)))

(defun arj/git-commit-setup ()
  (cl-case arj/magit-commit-or-reword
    (:commit (evil-insert-state))
    (:reword (evil-normal-state))))

(add-hook 'git-commit-setup-hook #'arj/git-commit-setup)
```

**Reply by tarsius** [reply]
You are welcome!
